### PR TITLE
Feat: JsxScope - Capture metrics only for files contained within local package

### LINKED ISSUE #127: JSX Scope - Revisit getSourceFiles logic
Currently the logic for `getSourceFile` of the JSX scope is collecting element for all files contained in the github project.
This could result in duplicate collection if the repository is a monorepo with multiple versions of the instrumented package being installed across different packages. Is there a way to refactor the `getSourceFiles` logic to get only the files from the specific package the instrumented package is installed in?

Brainstorm ideas:
Collect element for files that belong to the instrumented package (and version) ’s package.json

> find installers of instrumented package and version
>  
> `git ls-tree` with the package.json directory 
	> Collect elements only for these files only

### PR DESCRIPTION
**Pull request by francinelucca**
Closes #127 

-  Collects only JsxElement metrics for files that belong to the local package that installed the instrumented package
- Adds tests for new/corrected functionality

#### Changelog

**New**

- New `OpenTelemetryContext` src/main/core/open-telemetry-context.ts
- New `NoPackageDataFoundError` exception src/main/exceptions/no-package-data-found-error.ts
- New `NoInstallationFoundError` exception src/main/exceptions/no-installation-found-error.ts
- New `findRelevantSourceFiles`, `getTreePredecessor`, `getPackageTrees` and `getInstalledVersionPaths` functions in `JsxScope` - `src/main/scopes/jsx/jsx-scope.ts`
- Added more fixture test files
- New test files and test cases

**Changed**

- Offloaded `NpmScope`'s `findInstallingPackages` into a utility in `src/main/npm/find-installing-packages.ts` so that it can be reused
- Offloaded `findInstallingPackages`'s dependency tree logic onto new `getDependencyTree` standalone function
- Renamed JsxScope's `captureElementMetrics` function to `captureAllMetrics`
- Modified `captureAllMetrics` logic to use new `findRelevantSourceFiles` function to determine files that belong to the local package(s) that installed the instrumented package.
- Modified JsxScope's `removeIrrelevantImports` functionality to match imports package names exactly (not only startsWith).
- Replace `packageVersion` optional parameter in `findInstallersFromTree`, `findInstallingPackages`, and `findNestedDeps` with a custom filterFn that may or may not filter by version.

**Removed**
- `InitializeOpenTelemetry` in favor of new `OpenTelemetryContext`

#### Testing / reviewing

Unit tests should pass and code should pass comprehensive test

### CODE REVIEW

File: src/main/scopes/jsx/jsx-scope.ts [hunk 1]
```diff
@@ -58,28 +59,67 @@ export class JsxScope extends Scope {
...
+
+    const filterPromises = sourceFiles.map(async (f) => {
+      const prefix = await getDirectoryPrefix(path.dirname(f.fileName), this.logger)
+      const prefixPackageData = await getPackageData(prefix, this.root, this.logger)
+      const installingPackages = await findInstallingPackages(
+        this.cwd,
+        this.root,
+        instrumentedPackage,
+        this.logger
+      )
```
Comment (francinelucca) [inline src/main/scopes/jsx/jsx-scope.ts:110]: could this be outside of the map? so it isn't recomputed on every iteration?

File: src/main/scopes/npm/find-installing-packages.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+ * If no installers are found after this project directory is searched, an empty array is returned.
+ *
+ * @param cwd - Current working directory. This must be inside of the root directory. This is an
+ * absolute path.
+ * @param root - Root-most directory. This is an absolute path.
+ * @param packageData - The name and exact version of the package to search for.
+ * @param logger - Logger instance.
+ * @returns A possibly empty array of installing packages.
+ */
+export async function findInstallingPackages(
```
Comment (francinelucca) [inline src/main/scopes/npm/find-installing-packages.ts:31]: is there a difference between this file and src/main/core/find-installing-packages.ts?

File: src/test/scopes/jsx/__snapshots__/jsx-scope.e2e.test.ts.snap [hunk 3]
```diff
@@ -1,5 +1,78 @@
 // Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html
 
+exports[`class: JsxScope > findLocalPackages > correctly finds local packages 1`] = `
```
Comment (francinelucca) [inline src/test/scopes/jsx/__snapshots__/jsx-scope.e2e.test.ts.snap:3]: snapshpot needs to be updated

File: src/test/scopes/npm/find-installing-packages.test.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright IBM Corp. 2023, 2023
+ *
+ * This source code is licensed under the Apache-2.0 license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+import path from 'node:path'
```
Comment (francinelucca) [inline src/test/scopes/npm/find-installing-packages.test.ts:7]: duplicate src/test/core/__snapshots__/find-installing-packages.test.ts.snap?

**Review by jdharvey-ibm — APPROVED** [review]

File: src/main/scopes/npm/find-installing-packages.ts (hunk 2 above)
Comment (jdharvey-ibm) [reply src/main/scopes/npm/find-installing-packages.ts:31]: Oops, there should only be one of them. Just a merge bug. I'll remove the core one.

File: src/main/scopes/jsx/jsx-scope.ts (hunk 1 above)
Comment (jdharvey-ibm) [reply src/main/scopes/jsx/jsx-scope.ts:110]: Definitely! Good spot! Since this is now cached, it shouldn't slow things down to have it in the map, but I'll move it out since that's definitely how it should be.

File: src/test/scopes/jsx/__snapshots__/jsx-scope.e2e.test.ts.snap (hunk 3 above)
Comment (jdharvey-ibm) [reply src/test/scopes/jsx/__snapshots__/jsx-scope.e2e.test.ts.snap]: Done! Apparently my test:watch command and hitting `u` didn't remove the obsolete ones. Should be all good now though.

File: src/main/scopes/jsx/jsx-scope.ts [hunk 5]
```diff
@@ -59,28 +63,170 @@ export class JsxScope extends Scope {
...
+   * @param dependencyTree - The tree to search.
+   * @param pkgName - The tree to search.
+   * @returns A (possibly empty) array of paths.
+   */
+  getInstalledVersionPaths(dependencyTree: DependencyTree, pkgName: string) {
+    // find all versions, sort by shortest paths
+    const instrumentedInstallPaths = findNestedDeps(dependencyTree, pkgName, () => true).sort(
+      (a, b) => {
+        if (a.length === b.length) return 0
+        return a.length < b.length ? -1 : 1
```
Comment (jdharvey-ibm) [inline src/main/scopes/jsx/jsx-scope.ts:160]: shorthand for this could be:

`return a.length - b.length`

File: src/main/scopes/npm/get-dependency-tree.ts [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    )
+
+    const npmTree = JSON.parse(commandResult.stdout)
+
+    logger.traceExit('', 'getDependencyTree inner promise', npmTree)
+    return npmTree
+  }
+
+  // Allow this command to try and obtain results even if it exited with a total or partial
+  // error
```
Comment (jdharvey-ibm) [inline src/main/scopes/npm/get-dependency-tree.ts:63]: This comment should be up above line 49

File: src/test/__fixtures/projects/complex-nesting-thingy/package1/test.jsx [hunk 7]
```diff
@@ -0,0 +1,15 @@
+// @ts-nocheck
```
Comment (jdharvey-ibm) [inline src/test/__fixtures/projects/complex-nesting-thingy/package1/test.jsx:1]: I'm wondering if we can move some of this logic over to the "multiple instances of same dep" fixture instead. I feel like we have a tendency to use this complex fixture, but don't have a great handle on all of the edge cases it covers.

Comment (francinelucca) [reply src/test/__fixtures/projects/complex-nesting-thingy/package1/test.jsx:1]: https://github.com/ibm-telemetry/telemetry-js/issues/173

File: src/main/scopes/npm/get-dependency-tree.ts (hunk 6 above)
Comment (francinelucca) [reply src/main/scopes/npm/get-dependency-tree.ts:63]: https://github.com/ibm-telemetry/telemetry-js/issues/174

File: src/main/scopes/jsx/jsx-scope.ts (hunk 5 above)
Comment (jdharvey-ibm) [reply src/main/scopes/jsx/jsx-scope.ts:160]: #175

**Review by jdharvey-ibm — APPROVED** [review]
